app/code/Magento/InventorySales/Observer/SourceDeductionProcessor.php

```diff
@@ -0,0 +1,154 @@
+{
+    /**
+     * @var SearchCriteriaBuilderFactory
+     */
+    private $searchCriteriaBuilderFactory;
+
+    /**
+     * @var SourceItemRepositoryInterface
+     */
+    private $sourceItemRepository;
+
+    /**
+     * @var ReservationBuilderInterface
+     */
+    private $reservationBuilder;
+
+    /**
+     * @var AppendReservationsInterface
+     */
+    private $appendReservations;
+
+    /**
+     * @var SourceItemsSaveInterface
+     */
+    private $sourceItemsSave;
+
+    /**
+     * @var StockByWebsiteIdResolver
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * SourceDeductionProcessor constructor.
+     * @param SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory
+     * @param SourceItemRepositoryInterface $sourceItemRepository
+     * @param ReservationBuilderInterface $reservationBuilder
+     * @param AppendReservationsInterface $appendReservations
+     * @param SourceItemsSaveInterface $sourceItemsSave
+     * @param StockByWebsiteIdResolver $stockByWebsiteIdResolver
+     */
+    public function __construct(
+        SearchCriteriaBuilderFactory $searchCriteriaBuilderFactory,
+        SourceItemRepositoryInterface $sourceItemRepository,
+        ReservationBuilderInterface $reservationBuilder,
+        AppendReservationsInterface $appendReservations,
+        SourceItemsSaveInterface $sourceItemsSave,
+        StockByWebsiteIdResolver $stockByWebsiteIdResolver
+    ) {
+        $this->searchCriteriaBuilderFactory = $searchCriteriaBuilderFactory;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->reservationBuilder = $reservationBuilder;
+        $this->appendReservations = $appendReservations;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Shipment $shipment */
+        $shipment = $observer->getEvent()->getShipment();
+        if ($shipment->getOrigData('entity_id')) {
+            return $this;
+        }
+
+        $order = $shipment->getOrder();
+
+        // I'm not sure about websiteId
+        $websiteId = $order->getStore()->getWebsiteId();
+
+        foreach ($shipment->getItems() as $item) {
+            if (!$item->getSources()) {
+                continue;
+            }
+            foreach ($item->getSources() as $source) {
+                $source_code = $source['source_code'];
+                $qty = (float)$source['qty'];
+                $sourceItem = $this->getSourceItemByCodeAndSku($source_code, $item->getSku());
+                if (($sourceItem->getQuantity() - $qty) >= 0) {
+                    $sourceItem->setQuantity($sourceItem->getQuantity() - $qty);
+                    $this->sourceItemsSave->execute([$sourceItem]);
```

Comment:
why do we make SourceItem persistence at once, but don't collect all source items need to be persisted.
And save all of them at once. That should be more efficient